server/mediaserver.go

```diff
@@ -481,12 +287,88 @@ func endRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 		if monitor.Enabled {
 			monitor.LogStreamEndedEvent(cxn.nonce)
 		}
+		cxn.eof <- struct{}{}
 		delete(s.rtmpConnections, mid)
 
 		return nil
 	}
 }
 
+func (s *LivepeerServer) startSession(cxn *rtmpConnection) *BroadcastSession {
+
+	mid := rtmpManifestID(cxn.stream)
+	cpl := cxn.pl
+	var sess *BroadcastSession
+
+	// Connect to the orchestrator. If it fails, retry for as long
+	// as the RTMP stream is alive
+	broadcastFunc := func() error {
+		var err error
+		sess, err = selectOrchestrator(s.LivepeerNode, cpl)
+		if err == ErrDiscovery {
+			return nil // discovery disabled, don't retry
+		} else if err != nil {
+			// Should be logged upstream
+		}
+		s.connectionLock.RLock()
+		defer s.connectionLock.RUnlock()
+		_, active := s.rtmpConnections[mid]
+		if active {
+			return err
+		}
+		return nil // stop if inactive
+	}
+	expb := backoff.NewExponentialBackOff()
+	expb.MaxInterval = BroadcastRetry
+	expb.MaxElapsedTime = 0
+	backoff.Retry(broadcastFunc, expb)
```

Comment:
What is the point of using backoff here? If it doesn't succeed on the first run, results of any next run will be just lost. Probably it is just left from old code?

Reply:
Not sure that I understand the question. If acquiring a session doesn't succeed, it retries for as long as the RTMP client is connected.

Reply:
Yes, it retries it, and puts result in local var on line 306, and that result doesn't go anywhere - `startSession` is synchronous function that should return result (session) immediately.

Reply:
Just want to be sure we're looking at the same place -- the `sess` is declared outside of the `broadcastFunc` function that actually calls `selectOrchestrator` (so `sess` is non-local in this case).

https://github.com/livepeer/go-livepeer/pull/646/files/1a5f92d1c5902044bc33933cc9bf7cb8eeb07b72#diff-236217e7f473dfbde9134c6a391f3408R307

This is one of the small functional changes that was made during the refactoring; rather than assigning the `sess` to the `cxn` here as was done earlier, we do the assignment during `startSessionListener` in order to keep the state mutation somewhat self-contained to startSessionListener. Exact commit where this was done: https://github.com/livepeer/go-livepeer/pull/646/commits/bed60552e05f416c193a8d1f6d653a198cdb44b3

Reply:
Sorry, my mistake - I, after reading documentation on `backoff.Retry`, somehow thought that it operates asynchronously.